File: ios/Sources/IsRootPlugin/IsRootPlugin.swift
import Foundation
import Capacitor
#if canImport(UIKit)
import UIKit
#endif

@objc(IsRootPlugin)
public class IsRootPlugin: CAPPlugin, CAPBridgedPlugin {
    private let pluginVersion: String = "8.1.4"
    public let identifier = "IsRootPlugin"
    public let jsName = "IsRoot"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isRooted", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isRootedWithBusyBox", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "detectRootManagementApps", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "detectPotentiallyDangerousApps", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "detectTestKeys", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkForBusyBoxBinary", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkForSuBinary", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkSuExists", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkForRWPaths", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkForDangerousProps", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkForRootNative", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "detectRootCloakingApps", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isSelinuxFlagInEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isExistBuildTags", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "doesSuperuserApkExist", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isExistSUPath", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkDirPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkExecutingCommands", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkInstalledPackages", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkforOverTheAirCertificates", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isRunningOnEmulator", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "simpleCheckEmulator", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "simpleCheckSDKBF86", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "simpleCheckQRREFPH", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "simpleCheckBuild", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkGenymotion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkGeneric", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkGoogleSDK", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "togetDeviceInfo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isRootedWithEmulator", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isRootedWithBusyBoxWithEmulator", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let detector = IsRoot()

    @objc func isRooted(_ call: CAPPluginCall) {
        resolve(call, value: detector.isRooted())
    }

    @objc func isRootedWithBusyBox(_ call: CAPPluginCall) {
        resolve(call, value: detector.isRooted())
    }

    @objc func detectRootManagementApps(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func detectPotentiallyDangerousApps(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func detectTestKeys(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func checkForBusyBoxBinary(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func checkForSuBinary(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func checkSuExists(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func checkForRWPaths(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func checkForDangerousProps(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func checkForRootNative(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func detectRootCloakingApps(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func isSelinuxFlagInEnabled(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func isExistBuildTags(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func doesSuperuserApkExist(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func isExistSUPath(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func checkDirPermissions(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func checkExecutingCommands(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func checkInstalledPackages(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func checkforOverTheAirCertificates(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func isRunningOnEmulator(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func simpleCheckEmulator(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func simpleCheckSDKBF86(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func simpleCheckQRREFPH(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func simpleCheckBuild(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func checkGenymotion(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func checkGeneric(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func checkGoogleSDK(_ call: CAPPluginCall) {
        resolve(call, value: false)
    }

    @objc func togetDeviceInfo(_ call: CAPPluginCall) {
        #if canImport(UIKit)
        let device = UIDevice.current
        let info: [String: Any] = [
            "name": device.name,
            "model": device.model,
            "systemName": device.systemName,
            "systemVersion": device.systemVersion,
            "identifierForVendor": device.identifierForVendor?.uuidString ?? ""
        ]
        call.resolve(info)
        #else
        let processInfo = ProcessInfo.processInfo
        let info: [String: Any] = [
            "name": processInfo.hostName,
            "model": "",
            "systemName": processInfo.operatingSystemVersionString,
            "systemVersion": processInfo.operatingSystemVersionString,
            "identifierForVendor": ""
        ]
        call.resolve(info)
        #endif
    }

    @objc func isRootedWithEmulator(_ call: CAPPluginCall) {
        resolve(call, value: detector.isRooted())
    }

    @objc func isRootedWithBusyBoxWithEmulator(_ call: CAPPluginCall) {
        resolve(call, value: detector.isRooted())
    }

    private func resolve(_ call: CAPPluginCall, value: Bool) {
        call.resolve(["result": value])
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve(["version": self.pluginVersion])
    }

}

File: android/src/main/java/app/capgo/plugin/isroot/IsRootPlugin.java
package app.capgo.plugin.isroot;

import android.content.Context;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import org.json.JSONException;
import org.json.JSONObject;

@CapacitorPlugin(name = "IsRoot")
public class IsRootPlugin extends Plugin {

    private final String pluginVersion = "8.1.9";

    private final IsRoot implementation = new IsRoot();

    @PluginMethod
    public void isRooted(PluginCall call) {
        handleBoolean(call, () -> implementation.isRooted(requireContext()));
    }

    @PluginMethod
    public void isRootedWithBusyBox(PluginCall call) {
        handleBoolean(call, () -> implementation.isRootedWithBusyBox(requireContext()));
    }

    @PluginMethod
    public void detectRootManagementApps(PluginCall call) {
        handleBoolean(call, () -> implementation.detectRootManagementApps(requireContext()));
    }

    @PluginMethod
    public void detectPotentiallyDangerousApps(PluginCall call) {
        handleBoolean(call, () -> implementation.detectPotentiallyDangerousApps(requireContext()));
    }

    @PluginMethod
    public void detectTestKeys(PluginCall call) {
        handleBoolean(call, () -> implementation.detectTestKeys(requireContext()));
    }

    @PluginMethod
    public void checkForBusyBoxBinary(PluginCall call) {
        handleBoolean(call, () -> implementation.checkForBusyBoxBinary(requireContext()));
    }

    @PluginMethod
    public void checkForSuBinary(PluginCall call) {
        handleBoolean(call, () -> implementation.checkForSuBinary(requireContext()));
    }

    @PluginMethod
    public void checkSuExists(PluginCall call) {
        handleBoolean(call, () -> implementation.checkSuExists(requireContext()));
    }

    @PluginMethod
    public void checkForRWPaths(PluginCall call) {
        handleBoolean(call, () -> implementation.checkForRWPaths(requireContext()));
    }

    @PluginMethod
    public void checkForDangerousProps(PluginCall call) {
        handleBoolean(call, () -> implementation.checkForDangerousProps(requireContext()));
    }

    @PluginMethod
    public void checkForRootNative(PluginCall call) {
        handleBoolean(call, () -> implementation.checkForRootNative(requireContext()));
    }

    @PluginMethod
    public void detectRootCloakingApps(PluginCall call) {
        handleBoolean(call, () -> implementation.detectRootCloakingApps(requireContext()));
    }

    @PluginMethod
    public void isSelinuxFlagInEnabled(PluginCall call) {
        handleBoolean(call, implementation::isSelinuxFlagInEnabled);
    }

    @PluginMethod
    public void isExistBuildTags(PluginCall call) {
        handleBoolean(call, () -> implementation.runInternalCheck("isExistBuildTags", requireContext()));
    }

    @PluginMethod
    public void doesSuperuserApkExist(PluginCall call) {
        handleBoolean(call, () -> implementation.runInternalCheck("doesSuperuserApkExist", requireContext()));
    }

    @PluginMethod
    public void isExistSUPath(PluginCall call) {
        handleBoolean(call, () -> implementation.runInternalCheck("isExistSUPath", requireContext()));
    }

    @PluginMethod
    public void checkDirPermissions(PluginCall call) {
        handleBoolean(call, () -> implementation.runInternalCheck("checkDirPermissions", requireContext()));
    }

    @PluginMethod
    public void checkExecutingCommands(PluginCall call) {
        handleBoolean(call, () -> implementation.runInternalCheck("checkExecutingCommands", requireContext()));
    }

    @PluginMethod
    public void checkInstalledPackages(PluginCall call) {
        handleBoolean(call, () -> implementation.runInternalCheck("checkInstalledPackages", requireContext()));
    }

    @PluginMethod
    public void checkforOverTheAirCertificates(PluginCall call) {
        handleBoolean(call, () -> implementation.runInternalCheck("checkforOverTheAirCertificates", requireContext()));
    }

    @PluginMethod
    public void isRunningOnEmulator(PluginCall call) {
        handleBoolean(call, () -> implementation.runInternalCheck("isRunningOnEmulator", requireContext()));
    }

    @PluginMethod
    public void simpleCheckEmulator(PluginCall call) {
        handleBoolean(call, () -> implementation.runInternalCheck("simpleCheckEmulator", requireContext()));
    }

    @PluginMethod
    public void simpleCheckSDKBF86(PluginCall call) {
        handleBoolean(call, () -> implementation.runInternalCheck("simpleCheckSDKBF86", requireContext()));
    }

    @PluginMethod
    public void simpleCheckQRREFPH(PluginCall call) {
        handleBoolean(call, () -> implementation.runInternalCheck("simpleCheckQRREFPH", requireContext()));
    }

    @PluginMethod
    public void simpleCheckBuild(PluginCall call) {
        handleBoolean(call, () -> implementation.runInternalCheck("simpleCheckBuild", requireContext()));
    }

    @PluginMethod
    public void checkGenymotion(PluginCall call) {
        handleBoolean(call, () -> implementation.runInternalCheck("checkGenymotion", requireContext()));
    }

    @PluginMethod
    public void checkGeneric(PluginCall call) {
        handleBoolean(call, () -> implementation.runInternalCheck("checkGeneric", requireContext()));
    }

    @PluginMethod
    public void checkGoogleSDK(PluginCall call) {
        handleBoolean(call, () -> implementation.runInternalCheck("checkGoogleSDK", requireContext()));
    }

    @PluginMethod
    public void togetDeviceInfo(PluginCall call) {
        handleObject(call, () -> implementation.getDeviceInfo());
    }

    @PluginMethod
    public void isRootedWithEmulator(PluginCall call) {
        handleBoolean(call, () -> implementation.isRootedWithEmulator(requireContext()));
    }

    @PluginMethod
    public void isRootedWithBusyBoxWithEmulator(PluginCall call) {
        handleBoolean(call, () -> implementation.isRootedWithBusyBoxWithEmulator(requireContext()));
    }

    private Context requireContext() {
        Context context = getContext();
        if (context == null) {
            throw new IllegalStateException("Context is not available");
        }
        return context;
    }

    private void handleBoolean(PluginCall call, BooleanSupplierWithException task) {
        execute(() -> {
            try {
                boolean result = task.get();
                JSObject ret = new JSObject();
                ret.put("result", result);
                call.resolve(ret);
            } catch (JSONException ex) {
                call.reject("Unable to encode result", ex);
            } catch (Exception ex) {
                call.reject(ex.getMessage(), ex);
            }
        });
    }

    private void handleObject(PluginCall call, ObjectSupplierWithException task) {
        execute(() -> {
            try {
                Object result = task.get();
                if (result instanceof JSONObject) {
                    JSObject ret = JSObject.fromJSONObject((JSONObject) result);
                    call.resolve(ret);
                } else if (result instanceof JSObject) {
                    call.resolve((JSObject) result);
                } else {
                    JSObject ret = new JSObject();
                    ret.put("value", result);
                    call.resolve(ret);
                }
            } catch (JSONException ex) {
                call.reject("Unable to encode JSON", ex);
            } catch (Exception ex) {
                call.reject(ex.getMessage(), ex);
            }
        });
    }

    @FunctionalInterface
    private interface BooleanSupplierWithException {
        boolean get() throws Exception;
    }

    @FunctionalInterface
    private interface ObjectSupplierWithException {
        Object get() throws Exception;
    }

    @PluginMethod
    public void getPluginVersion(final PluginCall call) {
        try {
            final JSObject ret = new JSObject();
            ret.put("version", this.pluginVersion);
            call.resolve(ret);
        } catch (final Exception e) {
            call.reject("Could not get plugin version", e);
        }
    }
}
